hibernate-orm: hbm-xml: Cannot serialise field 'scope' on object 'javax.xml.bind.JAXBElement@7d024b28' as the property is read only

### Describe the bug

If we have any filter-def tag as part of hbm.xml than we get below exception while building project.

```
java.lang.RuntimeException: Failed to record call to method public io.quarkus.arc.runtime.BeanContainerListener io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder.initMetadata(java.util.List,org.hibernate.boot.archive.scan.spi.Scanner,java.util.Collection,io.quarkus.hibernate.orm.runtime.proxies.PreGeneratedProxies)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:509)
	at io.quarkus.deployment.steps.MainClassBuildStep.writeRecordedBytecode(MainClassBuildStep.java:436)
	at io.quarkus.deployment.steps.MainClassBuildStep.build(MainClassBuildStep.java:168)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.lang.RuntimeException: Failed to substitute io.quarkus.hibernate.orm.runtime.boot.QuarkusPersistenceUnitDefinition@79b350e1
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:672)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstance(BytecodeRecorderImpl.java:610)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1096)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:938)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstance(BytecodeRecorderImpl.java:610)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:504)
	... 13 more
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: Cannot serialise field 'scope' on object 'javax.xml.bind.JAXBElement@7d024b28' as the property is read only
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1393)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:938)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstance(BytecodeRecorderImpl.java:610)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:652)
	... 18 more
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: Cannot serialise field 'scope' on object 'javax.xml.bind.JAXBElement@7d024b28' as the property is read only
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1393)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:938)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstance(BytecodeRecorderImpl.java:610)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1096)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:938)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstance(BytecodeRecorderImpl.java:610)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1356)
	... 21 more
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: Cannot serialise field 'scope' on object 'javax.xml.bind.JAXBElement@7d024b28' as the property is read only
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1312)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:938)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstance(BytecodeRecorderImpl.java:610)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1356)
	... 27 more
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: Cannot serialise field 'scope' on object 'javax.xml.bind.JAXBElement@7d024b28' as the property is read only
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1312)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:938)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstance(BytecodeRecorderImpl.java:610)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1232)
	... 30 more
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: Cannot serialise field 'scope' on object 'javax.xml.bind.JAXBElement@7d024b28' as the property is read only
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1312)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstanceImpl(BytecodeRecorderImpl.java:938)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadObjectInstance(BytecodeRecorderImpl.java:610)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1232)
	... 33 more
Caused by: java.lang.RuntimeException: Cannot serialise field 'scope' on object 'javax.xml.bind.JAXBElement@7d024b28' as the property is read only
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1303)
	... 36 more

```

### Expected behavior

Project should build successfully.

### Actual behavior

Getting Exception.

### How to Reproduce?

Steps to Reproduce:

1. Copy following contents in extensions/hibernate-orm/deployment/src/test/resources/META-INF/hbm-component.xml
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE hibernate-mapping PUBLIC
        "-//Hibernate/Hibernate Mapping DTD 3.0//EN"
        "http://www.hibernate.org/dtd/hibernate-mapping-3.0.dtd">
<hibernate-mapping>
    <class name="io.quarkus.hibernate.orm.xml.hbm.NonAnnotatedComponentUsingEntity">
        <id name="id">
            <generator class="identity"/>
        </id>
        <property name="groupId" type="long">
            <column name="GroupId" not-null="true" default="0"/>
        </property>
        <property name="authId" column="AuthId" />
        <component name="value" class="io.quarkus.hibernate.orm.xml.hbm.NonAnnotatedComponent" lazy="false">
            <property name="name">
                <column name="thename"/>
            </property>
        </component>
        <filter name="filter-name" condition="AuthId in (select gm.MemberId from group_members where groupId in (:groupIds)"/>
    </class>
    <filter-def name="filter-name" condition="AuthId in (select gm.MemberId from group_members where groupId in (:groupIds)">
        <filter-param name="groupIds" type="long"/>
    </filter-def>
</hibernate-mapping>
```
2. Run Test: HbmXmlComponentTest

### Output of `uname -a` or `ver`

Linux ubuntu 4.15.0-159-generic #167-Ubuntu SMP Tue Sep 21 08:55:05 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.11" 2021-04-20 OpenJDK Runtime Environment (build 11.0.11+9-Ubuntu-0ubuntu2.18.04) OpenJDK 64-Bit Server VM (build 11.0.11+9-Ubuntu-0ubuntu2.18.04, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.6.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)

### Additional information

_No response_

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
@@ -41,6 +41,8 @@ import javax.persistence.SharedCacheMode;
 import javax.persistence.ValidationMode;
 import javax.persistence.spi.PersistenceUnitTransactionType;
 import javax.transaction.TransactionManager;
+import javax.xml.bind.JAXBElement;
+import javax.xml.namespace.QName;
 
 import org.hibernate.MultiTenancyStrategy;
 import org.hibernate.boot.archive.scan.spi.ClassDescriptor;
@@ -124,6 +126,8 @@ import io.quarkus.hibernate.orm.runtime.RequestScopedSessionHolder;
 import io.quarkus.hibernate.orm.runtime.TransactionSessions;
 import io.quarkus.hibernate.orm.runtime.boot.QuarkusPersistenceUnitDefinition;
 import io.quarkus.hibernate.orm.runtime.boot.scan.QuarkusScanner;
+import io.quarkus.hibernate.orm.runtime.boot.xml.JAXBElementSubstitution;
+import io.quarkus.hibernate.orm.runtime.boot.xml.QNameSubstitution;
 import io.quarkus.hibernate.orm.runtime.boot.xml.RecordableXmlMapping;
 import io.quarkus.hibernate.orm.runtime.cdi.QuarkusArcBeanContainer;
 import io.quarkus.hibernate.orm.runtime.devconsole.HibernateOrmDevConsoleCreateDDLSupplier;
@@ -602,6 +606,16 @@ public final class HibernateOrmProcessor {
         recorderContext.registerSubstitution(QuarkusPersistenceUnitDefinition.class,
                 QuarkusPersistenceUnitDefinition.Serialized.class,
                 QuarkusPersistenceUnitDefinition.Substitution.class);
+
+        if (hasXmlMappings(persistenceUnitDescriptorBuildItems)) {
+            //Make it possible to record JAXBElement as bytecode:
+            recorderContext.registerSubstitution(JAXBElement.class,
+                    JAXBElementSubstitution.Serialized.class,
+                    JAXBElementSubstitution.class);
+            recorderContext.registerSubstitution(QName.class,
+                    QNameSubstitution.Serialized.class,
+                    QNameSubstitution.class);
+        }
 
         beanContainerListener
                 .produce(new BeanContainerListenerBuildItem(

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/xml/hbm/HbmXmlFilterDefTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/xml/hbm/HbmXmlFilterDefTest.java
@@ -1,0 +1,61 @@
+package io.quarkus.hibernate.orm.xml.hbm;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Collections;
+
+import javax.inject.Inject;
+import javax.transaction.Transactional;
+
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.hibernate.orm.SmokeTestUtils;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class HbmXmlFilterDefTest {
+    @RegisterExtension
+    final static QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .withApplicationRoot(jar -> jar
+                    .addClass(SmokeTestUtils.class)
+                    .addClass(NonAnnotatedEntity.class)
+                    .addAsResource("META-INF/hbm-filterdef.xml", "my-hbm.xml"))
+            .withConfigurationResource("application.properties")
+            .overrideConfigKey("quarkus.hibernate-orm.mapping-files", "my-hbm.xml")
+            .overrideConfigKey("quarkus.hibernate-orm.log.sql", "true");
+
+    @Inject
+    SessionFactory sessionFactory;
+
+    @Inject
+    Session session;
+
+    @Test
+    @Transactional
+    public void hbmXmlTakenIntoAccount() {
+        assertThat(sessionFactory.getDefinedFilterNames())
+                .contains("idFilter");
+    }
+
+    @Test
+    @Transactional
+    public void smokeTest() {
+        NonAnnotatedEntity firstEntity = new NonAnnotatedEntity("first");
+        session.persist(firstEntity);
+        NonAnnotatedEntity secondEntity = new NonAnnotatedEntity("second");
+        session.persist(secondEntity);
+        session.flush();
+
+        assertThat(session.createQuery("select e from " + NonAnnotatedEntity.class.getName() + " e",
+                NonAnnotatedEntity.class).list())
+                        .hasSize(2);
+        session.enableFilter("idFilter")
+                .setParameterList("ids", Collections.singletonList(firstEntity.getId()));
+        assertThat(session.createQuery("select e from " + NonAnnotatedEntity.class.getName() + " e",
+                NonAnnotatedEntity.class).list())
+                        .hasSize(1);
+    }
+
+}

--- a/extensions/hibernate-orm/deployment/src/test/resources/META-INF/hbm-filterdef.xml
+++ b/extensions/hibernate-orm/deployment/src/test/resources/META-INF/hbm-filterdef.xml
@@ -1,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE hibernate-mapping PUBLIC
+        "-//Hibernate/Hibernate Mapping DTD 3.0//EN"
+        "http://www.hibernate.org/dtd/hibernate-mapping-3.0.dtd">
+<hibernate-mapping>
+    <class name="io.quarkus.hibernate.orm.xml.hbm.NonAnnotatedEntity">
+        <id name="id">
+            <generator class="identity"/>
+        </id>
+        <property name="name" column="thename"/>
+        <filter name="idFilter" condition="id in (:ids)"/>
+    </class>
+    <filter-def name="idFilter">
+        <filter-param name="ids" type="long"/>
+    </filter-def>
+</hibernate-mapping>

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/xml/JAXBElementSubstitution.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/xml/JAXBElementSubstitution.java
@@ -1,0 +1,37 @@
+package io.quarkus.hibernate.orm.runtime.boot.xml;
+
+import javax.xml.bind.JAXBElement;
+import javax.xml.namespace.QName;
+
+import io.quarkus.runtime.ObjectSubstitution;
+import io.quarkus.runtime.annotations.RecordableConstructor;
+
+@SuppressWarnings({ "unchecked", "rawtypes" })
+public class JAXBElementSubstitution implements ObjectSubstitution<JAXBElement, JAXBElementSubstitution.Serialized> {
+
+    @Override
+    public Serialized serialize(JAXBElement obj) {
+        return new Serialized(obj.getName(), obj.getDeclaredType(), obj.getScope(), obj.getValue());
+    }
+
+    @Override
+    public JAXBElement deserialize(Serialized obj) {
+        return new JAXBElement(obj.name, obj.declaredType, obj.scope, obj.value);
+    }
+
+    public static class Serialized {
+        public final QName name;
+        public final Class<?> declaredType;
+        public final Class<?> scope;
+        public final Object value;
+
+        @RecordableConstructor
+        public Serialized(QName name, Class<?> declaredType, Class<?> scope, Object value) {
+            this.name = name;
+            this.declaredType = declaredType;
+            this.scope = scope;
+            this.value = value;
+        }
+    }
+
+}

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/xml/QNameSubstitution.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/xml/QNameSubstitution.java
@@ -1,0 +1,33 @@
+package io.quarkus.hibernate.orm.runtime.boot.xml;
+
+import javax.xml.namespace.QName;
+
+import io.quarkus.runtime.ObjectSubstitution;
+import io.quarkus.runtime.annotations.RecordableConstructor;
+
+public class QNameSubstitution implements ObjectSubstitution<QName, QNameSubstitution.Serialized> {
+
+    @Override
+    public Serialized serialize(QName obj) {
+        return new Serialized(obj.getNamespaceURI(), obj.getLocalPart(), obj.getPrefix());
+    }
+
+    @Override
+    public QName deserialize(Serialized obj) {
+        return new QName(obj.namespaceURI, obj.localPart, obj.prefix);
+    }
+
+    public static class Serialized {
+        public final String namespaceURI;
+        public final String localPart;
+        public final String prefix;
+
+        @RecordableConstructor
+        public Serialized(String namespaceURI, String localPart, String prefix) {
+            this.namespaceURI = namespaceURI;
+            this.localPart = localPart;
+            this.prefix = prefix;
+        }
+    }
+
+}

